### PR TITLE
feat(sdk): bring back auth

### PR DESCRIPTION
**Pull request by ji-yaqi**
Fixes https://github.com/kubeflow/pipelines/issues/7466

### CODE REVIEW

File: sdk/python/kfp/client/set_volume_credentials.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+import os
+import logging
+
+from kubernetes.client import configuration
+
+from kfp import client
+
+
+class ServiceAccountTokenVolumeCredentials(client.TokenCredentialsBase):
```
Comment (chensun) [inline sdk/python/kfp/client/set_volume_credentials.py:23]: Need to update the import reference for this one:
https://github.com/kubeflow/pipelines/blob/dfc85b0a0d7f6e59bb3b1ced03d51a24a065b471/sdk/python/kfp/client/client.py#L341

File: sdk/python/kfp/client/__init__.py [hunk 2]
```diff
@@ -13,3 +13,9 @@
 # limitations under the License.
 
 from kfp.client.client import Client
+
+from kfp.client.token_credentials_base import TokenCredentialsBase, read_token_from_file
```
Comment (chensun) [inline sdk/python/kfp/client/__init__.py:17]: It looks to me that `read_token_from_file` doesn't need to be exposed from `kfp.client`.

File: sdk/python/kfp/client/set_volume_credentials.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import logging
+
+from kubernetes.client import configuration
+
+from kfp import client
```
Comment (chensun) [inline sdk/python/kfp/client/set_volume_credentials.py:20]: Would this import give you any error as the file in inside the same module?

**Review by connor-mccarthy — COMMENTED** [review]
/lgtm

nit: It might be nice to type-annotate this code before introducing to v2. WDYT?

File: sdk/python/kfp/client/set_volume_credentials.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    If the environment variable is also empty, it falls back to the path
+    specified in ``client.KF_PIPELINES_SA_TOKEN_PATH``.
+
+    This method of authentication is meant for use inside a Kubernetes cluster.
+
+    Relevant documentation:
+    https://kubernetes.io/docs/tasks/configure-pod-container/configure-service-account/#service-account-token-volume-projection
+    """
+
+    def __init__(self, path=None):
```
Comment (chensun) [inline sdk/python/kfp/client/set_volume_credentials.py:41]: nit: not your original code, but we should bring this up to our code style standard as we bring it back. So can you please add type annotation and docstrings for public interface?

File: sdk/python/kfp/client/set_volume_credentials.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        try:
+            token = client.read_token_from_file(self._token_path)
+        except OSError as e:
+            logging.error("Failed to read a token from file '%s' (%s).",
+                          self._token_path, str(e))
+            raise
+        return token
+
+    def refresh_api_key_hook(self, config: configuration.Configuration):
+        """Refresh the api key.
```
Comment (chensun) [inline sdk/python/kfp/client/set_volume_credentials.py:57]: nit: Refreshes

https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: sdk/python/kfp/client/token_credentials_base.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+                The configuration object that the client uses.
+
+                The Configuration object of the kubernetes client's is the same
+                with kfp_server_api.configuration.Configuration.
+        """
+        raise NotImplementedError()
+
+
+def read_token_from_file(path=None):
+    """Read a token found in some file."""
```
Comment (chensun) [inline sdk/python/kfp/client/token_credentials_base.py:43]: nit: Reads. Also the type annotation for `path`.

File: sdk/python/kfp/client/set_volume_credentials.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        self._token_path = (
+            path or os.getenv(client.KF_PIPELINES_SA_TOKEN_ENV) or
+            client.KF_PIPELINES_SA_TOKEN_PATH)
+
+    def _get_token(self):
+        token = None
+        try:
+            token = client.read_token_from_file(self._token_path)
+        except OSError as e:
+            logging.error("Failed to read a token from file '%s' (%s).",
```
Comment (connor-mccarthy) [inline sdk/python/kfp/client/set_volume_credentials.py:51]: What do we think about f-strings v other string formatting styles, @ji-yaqi @chensun?

The Style Guide doesn't state a preference [here](https://google.github.io/styleguide/pyguide.html#310-strings). I find f-strings to be the most readable string interpolation method, but either way would like to be consistent :)

**Comment by xin-hao-awx** [comment]
is this ready to be merged?

**Comment by connor-mccarthy** [comment]
This PR is completed by #7789. Closing.

cc: @xin-hao-awx

**Pull request closed without merging**